### PR TITLE
vm-virtio: Add snapshot/restore support to Vsock

### PR DESCRIPTION
**Pull request by sboeuf**
This pull request allows to snapshot and restore a VM with `virtio-vsock` devices attached.
As mentioned in the `vsock(7)` documentation:
```
Sockets are affected by live migration of virtual machines.  Con‐
nected SOCK_STREAM sockets become disconnected when the virtual
machine migrates to a new host.  Applications must reconnect when
this happens.
```
So we cannot expect a connection to survive over a snapshot/restore operation.

### CODE REVIEW

File: vm-virtio/src/vsock/device.rs [hunk 1]
```diff
@@ -394,6 +398,12 @@ pub struct Vsock<B: VsockBackend> {
     path: PathBuf,
 }
 
+#[derive(Serialize, Deserialize)]
+pub struct VsockState {
+    pub avail_features: u64,
+    pub acked_features: u64,
```
Comment (rbradford) [inline vm-virtio/src/vsock/device.rs:404]: Maybe we need something like a VirtioState to handle things like this.

File: vmm/src/device_manager.rs [hunk 2]
```diff
@@ -1982,6 +1985,12 @@ impl DeviceManager {
             .sock
             .to_str()
             .ok_or(DeviceManagerError::CreateVsockConvertPath)?;
+
+        // Make sure the socket path does not exist before we try to create it.
+        if Path::new(socket_path).exists() {
```
Comment (rbradford) [inline vmm/src/device_manager.rs:1990]: I'm sure it's unlinked on shutdown so i'm not sure why it would be there unless something has gone wrong?

Comment (sboeuf) [reply vmm/src/device_manager.rs:1990]: If the VM is simply killed with SIGKILL, it won't go through the Drop implementation.

File: vm-virtio/src/vsock/device.rs (hunk 1 above)
Comment (sboeuf) [reply vm-virtio/src/vsock/device.rs:404]: Yes I'll open an issue to factorize every virtio device state through a common VirtioState structure.

File: vmm/src/device_manager.rs (hunk 2 above)
Comment (sboeuf) [reply vmm/src/device_manager.rs]: I removed it!

File: vm-virtio/src/vsock/device.rs (hunk 1 above)
Comment (sboeuf) [reply vm-virtio/src/vsock/device.rs:404]: https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1211

**Pull request merged**